### PR TITLE
[MRG] Update surface functions

### PR DESCRIPTION
**Pull request by NicolasGensollen**
This PR attempts to update the surface plotting functions to comply with the new `Surface` object introduced in #2672, in order to obtain a more consistent API. See also #2681.

Assuming we have defined a surface like this:

```python
surf = Surface([mesh, map])
```

Usage:

- `plot_surf(surf_mesh, surf_map=None,...)` ==> `plot_surf(surf,...)` (map is optional)
- `plot_surf_contours(surf_mesh, roi_map,...)` ==> `plot_surf_contours(surf,...)` (map is required)
- `plot_surf_stat_map(surf_mesh, stat_map,...)` ==> `plot_surf_stat_map(surf,...)` (map is required)
- `plot_surf_roi(surf_mesh, roi_map,...)` ==> `plot_surf_roi(surf,...)` (map is required)
- `plot_img_on_surf(stat_map, surf_mesh='fsaverage5',...)` (map and mesh are reversed and special values can be passed to `surf_mesh`). Untouched so far, will probably be handled in a different way.

Passing a mesh and a map as before is still possible but gives a `FutureWarning` saying that this usage is deprecated and will be removed in release 0.9.

Passing a surface object gives a `UserWarning` saying that `surf.data` will be used as the surface map instead of what was potentially provided through `surf_map`, `roi_map`, or `stat_map`.

*Note: Although this is WIP, reviews/suggestions/ideas are welcome since there might be better ways to do this!*

### CODE REVIEW

File: nilearn/plotting/surf_plotting.py
Comment (tsalo) [inline nilearn/plotting/surf_plotting.py:32]:
```suggestion
def _deprecate_separate_mesh_data(func, argument):
```

Just a minor typo

**Review by bthirion — APPROVED** [review]
LGTM, thx !

File: examples/01_plotting/plot_surf_stat_map.py
Comment (bthirion) [inline examples/01_plotting/plot_surf_stat_map.py:157]: it begs the question whether bg_map should/could also be handled as a Surface ?

**Review by bthirion — COMMENTED** [review]
This probably deserves a whasnew entry, but LGTM otherwise.
Thx for that one !

**Comment by bthirion** [comment]
This LGTM. I just left a comment on the bg texture.

**Comment by NicolasGensollen** [comment]
> I just left a comment on the bg texture.

Sorry, I missed it. I am not sure what the best option is. 
One solution would be to change the texture `bg_map` parameter of type array to a "standalone" surface `bg_surf` parameter of type Surface, meaning that the background surface would hold both its mesh and background texture (you could basically do `plot_surf(background)` and it would work just fine).
Since the surface parameter is also a Surface, it also has a mesh such that we would need to handle situations where the user provide surface and background on different meshes (raise an error or default to surface mesh rather than background's...). 
The code would then look like this:

```python
import numpy as np
from nilearn.datasets import fetch_surf_fsaverage
from nilearn.surface import load_surface
from nilearn.plotting import plot_surf_stat_map

fsaverage = fetch_surf_fsaverage()
# Use a surface with random texture as stat map 
texture = np.random.random(10242)
surf = load_surface((fsaverage.infl_left,
                     texture))
# Define the background as a "standalone" surface object
background = load_surface((fsaverage.infl_left,
                           fsaverage.sulc_left))
# plot stat map
plot_surf_stat_map(surf,
                   bg_surf=background,
                   bg_on_data=True)
```

One of the reason we introduced `Surface` and `Mesh` objects is to avoid having untied textures, coordinates, and faces arrays in the code which makes me think that we should indeed provide a surface background rather than a simple texture although some information is redundant. 

Note that a much simpler solution would be to allow Surface objects with multiple textures as we discussed during the sprint. For example, we could provide a dictionary of `texture_label: texture_array` or something alike. 
This could for example look like:

```python
class Surface:
    def __init__(self, mesh, textures):
        self.mesh = mesh
        self._textures = textures
    def texture(self, label=""):
        if label in self._textures:
            return self._textures[label]
        else:
            return None

s = Surface(mesh=fsaverage.infl_left, 
            textures={'background': fsaverage.sulc_left,
                      'stat_map': texture})

# You can then do the following:
s.mesh
s.texture('background')
s.texture('stat_map')
# would be clever enough to plot both map and 
# background is there are matching keys...
plot_surf_stat_map(s)
```

**Comment by bthirion** [comment]
> Since the surface parameter is also a Surface, it also has a mesh such that we would need to handle situations where the user provide surface and background on different meshes (raise an error or default to surface mesh rather than background's...).

Note that there are cases where two meshes are different , yet in one-to-one correspondence, which is quite handy to compare them.  Freesurfer uses is quite often (e.g. pial and white mesh, white and inflated,  fsaverage and spherical, etc). You don't want to raise an error for such cases.
I'm a bit uncertain what to do, as having different textures called through keywords makes the API more complicated (and is reminicent of Gifti's intents).  I think I still prefer to have the texture as a separate surface, but this is not a strong opinion.

File: nilearn/plotting/html_surface.py
Comment (bthirion) [inline nilearn/plotting/html_surface.py:253]: Need to change the version

File: nilearn/plotting/surf_plotting.py
Comment (bthirion) [inline nilearn/plotting/surf_plotting.py:106]: Version to be chnaged here too.

**Comment by bthirion** [comment]
@alexisthual I'd love to have your opinion on that one.
Best,

**Comment by alexisthual** [comment]
Very nice work and ideas!

I haven't spent as much time as you working on this but here are some thoughts.

My opinion as a user is that I want to be free to manipulate `Texture` and `Mesh` objects independently and have them constitute a `Surface` object on the fly when needed. The current PR seems to do a very good job at this! To me, `bg_map` expecting a `Texture` object (and/or eventually a `Surface` whose `Mesh` would simply be ignored) seems natural.

I wouldn't allow a single `Surface` object to have several `Texture` objects available through keys (as a user, I would most likely store these textures in a custom dict I would handle myself). However, I feel like the broader underlying question has to do with the merging of several textures. To me, a user could want to merge several `Texture` objects into a single `Surface` object, and choose how to aggregate them. It could look something like this:

```python
fsaverage = fetch_surf_fsaverage()
texture = np.random.random(10242)
# create a surface whose texture is the point-wise product of two textures
surf = load_surface(
  (
    fsaverage.infl_left,
    [texture, fsaverage.sulc_left]
  ),
  agg="product" # could be "add", "max", etc.
)

plot_surf_stat_map(surf)
```

which, in most cases maybe, could also simply be achieved with `numpy`

```python
surf = load_surface(
  (
    fsaverage.infl_left,
    np.multiply(texture, fsaverage.sulc_left)
  )
)
```

meaning that this solution might be interesting only if there is a need for some complex aggregators.

One final idea would be to combine `Surface`s and/or `Texture`s using operators, allowing for a very concise syntax:

```python
fsaverage = fetch_surf_fsaverage()
texture = np.random.random(10242)
surf = load_surface((fsaverage.infl_left,
                     texture))

# plot point-wise product of surface and sulc texture
plot_surf(surf * fsaverage.sulc_left)
```

If any of this makes sense, it probably calls for a new PR :smile:

**Comment by jeromedockes** [comment]
> Besides, `bg_map` and `bg_on_data` didn't feel clear to me the first time I used them (why set `bg_on_data=True` if I already set a `bg_map`?), so maybe there is room for improvement.

I agree



> However, I feel like the broader underlying question has to do with the merging of several textures.

The main issue is when data from both hemispheres needs to be handled as a single unit, eg to be fed to a decoding pipeline (which I think would correspond to aggregating by concatenation in your description, but there needs to be an inverse transform)



> To me, `bg_map` is just a way to perform a product of textures before plotting it

yes but that's an implementation detail, we don't want the user to be aware of that or take care of the blending of surface map and background



> One final idea would be to combine `Surface`s and/or `Texture`s using operators, allowing for a very concise syntax:

this kind of syntactic sugar often leads to less readable code in my experience. I would prefer to restrict the overloading of mathematical operators to simple objects that truly behave like numbers or matrices

**Comment by bthirion** [comment]
Tx for the discussion.
One aspect is that there are different conventions for bg, which is plotted in grey scale vs the main texture, which gets a more flashy colorbar.

@jeromedockes I don't see how we handle the case with two meshes/textures for two hemispheres ?

**Comment by jeromedockes** [comment]
> @jeromedockes I don't see how we handle the case with two meshes/textures for two hemispheres ?

we don't. what I meant was that the issue discussed during the sprint had to do with Surfaces covering a whole brain, rather than multiple textures for a single mesh. given that this still isn't solved (ongoing discussion in [nibabel issue](https://github.com/nipy/nibabel/issues/936) ) I suggest leaving aside the multiple texture question for now.

for the bg_img, I also think it would be more consistent for it to be a Surface

**Comment by bthirion** [comment]
@NicolasGensollen could we have a quick discussion over that one ? Is it simply waiting for a second review ?

**Comment by bthirion** [comment]
Thx. We should at least discuss it to see how we coordinate with nibabel.

**Comment by ymzayek** [comment]
I think this is worth finishing (and then also working on #2424). Even though work is underway on the Coordinate Image API, I think it will be a while before we can adopt it even if it's merged soon considering waiting for a release and supporting older versions of nibabel. @NicolasGensollen and @jeromedockes WDYT? Do you foresee major issues in the future to convert to the new API after developing surface support a bit using our Surface object?

**Comment by ymzayek** [comment]
Can we close this since now we are working on #3856?

**Comment by Remi-Gau** [comment]
I would say yes

**Comment by Remi-Gau** [comment]
Turning this one to draft until we make more decisions on the new surface code is in.

**Comment by Remi-Gau** [comment]
Closing as this will be replaced by other PRs in the devlopment of surface based analysis.

Also the mount of merge conflicts is making this one almost impossible bring back.

**Comment by jeromedockes** [comment]
also this one was for the old surface api which has been abandoned

**Pull request closed without merging**
